File: src/Systems/Go/chain_of_responsibility.go
package main

import (
	"fmt"
	"strings"
)

type Handler struct {
	name      string
	maxAmount int
	next      *Handler
}

func (h *Handler) setNext(next *Handler) *Handler {
	h.next = next
	return next
}

func (h *Handler) handle(amount int, visited *[]string) string {
	*visited = append(*visited, h.name)
	if h.maxAmount < 0 || amount <= h.maxAmount {
		return h.name
	}
	if h.next == nil {
		panic("No handler accepted the request.")
	}
	return h.next.handle(amount, visited)
}

func main() {
	faq := &Handler{name: "faq", maxAmount: 50}
	billing := &Handler{name: "billing", maxAmount: 500}
	escalation := &Handler{name: "escalation", maxAmount: -1}
	faq.setNext(billing).setNext(escalation)

	visited := []string{}
	handled := faq.handle(250, &visited)
	fmt.Printf("visited=%s;handled=%s;result=refund(250)\n", strings.Join(visited, ">"), handled)
}

File: src/Systems/Rust/chain_of_responsibility.rs
struct Handler {
    name: &'static str,
    max_amount: Option<i32>,
    next: Option<Box<Handler>>,
}

impl Handler {
    fn new(name: &'static str, max_amount: Option<i32>, next: Option<Box<Handler>>) -> Self {
        Self {
            name,
            max_amount,
            next,
        }
    }

    fn handle<'a>(&'a self, amount: i32, visited: &mut Vec<&'a str>) -> &'a str {
        visited.push(self.name);
        if self.max_amount.is_none_or(|max| amount <= max) {
            return self.name;
        }

        self.next
            .as_deref()
            .expect("No handler accepted the request.")
            .handle(amount, visited)
    }
}

fn main() {
    let chain = Handler::new(
        "faq",
        Some(50),
        Some(Box::new(Handler::new(
            "billing",
            Some(500),
            Some(Box::new(Handler::new("escalation", None, None))),
        ))),
    );

    let mut visited = Vec::new();
    let handled = chain.handle(250, &mut visited);
    println!(
        "visited={};handled={};result=refund(250)",
        visited.join(">"),
        handled
    );
}
